src/controller/scan/base_controller.go

```diff
@@ -233,78 +266,116 @@ func (bc *basicController) Scan(ctx context.Context, artifact *ar.Artifact, opti
 	return nil
 }
 
-func (bc *basicController) makeReportPlaceholder(ctx context.Context, r *scanner.Registration, art *ar.Artifact, options ...Option) (string, []string, error) {
-	trackID, err := bc.uuid()
+func (bc *basicController) ScanAll(ctx context.Context, trigger string, async bool) (int64, error) {
+	extraAttrs := map[string]interface{}{}
+	executionID, err := bc.execMgr.Create(ctx, job.ImageScanAllJob, 0, trigger, extraAttrs)
 	if err != nil {
-		return "", nil, errors.Wrap(err, "scan controller: scan")
+		return 0, err
 	}
 
-	// Parse options
-	ops, err := parseOptions(options...)
-	if err != nil {
-		return "", nil, errors.Wrap(err, "scan controller: scan")
+	if async {
+		go bc.startScanAll(bc.makeCtx(), executionID, async)
+	} else {
+		if err := bc.startScanAll(ctx, executionID, async); err != nil {
+			return 0, err
+		}
 	}
 
-	create := func(ctx context.Context, digest, registrationUUID, mimeType, trackID string, status job.Status) error {
-		reportPlaceholder := &scan.Report{
-			Digest:           digest,
-			RegistrationUUID: registrationUUID,
-			Status:           status.String(),
-			StatusCode:       status.Code(),
-			TrackID:          trackID,
-			MimeType:         mimeType,
-		}
-		// Set requester if it is specified
-		if len(ops.Requester) > 0 {
-			reportPlaceholder.Requester = ops.Requester
-		} else {
-			// Use the trackID as the requester
-			reportPlaceholder.Requester = trackID
-		}
+	return executionID, nil
+}
 
-		_, e := bc.manager.Create(reportPlaceholder)
-		return e
+func (bc *basicController) startScanAll(ctx context.Context, executionID int64, async bool) error {
+	if async {
+		// if async, this method is running in another goroutine ensure the execution exists in db
+		err := lib.RetryUntil(func() error {
```

Comment:
Why we need this? Shouldn't the execution record already be there as you have called `bc.execMgr.Create`?

Reply:
When startScanAll in another goroutine, the API may not be finished but the `startScanAll` be finished. In this case `bc.execMgr.MarkDone(ctx, executionID, "no task found")` will failed because the execution not found in db.